File: src/test/java/org/veupathdb/lib/blast/BlastWithIPGListContractTest.java
package org.veupathdb.lib.blast;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.veupathdb.lib.blast.consts.Flag;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("BlastWithIPGList")
abstract class BlastWithIPGListContractTest extends BlastWithListsContractTest
{
  private final ObjectMapper json = new ObjectMapper();

  @Override
  abstract BlastWithIPGList newConfig();

  @Override
  abstract Class<? extends BlastWithIPGList> configClass();

  @Nested
  @DisplayName(Flag.IPGList)
  class IPGList1
  {
    @Test
    @DisplayName("serialization")
    void test1() throws Exception {
      var tgt = newConfig();
      tgt.setIPGList("hi");

      assertEquals("{\"-ipglist\":\"hi\"}", json.writeValueAsString(tgt));
    }

    @Test
    @DisplayName("deserialization")
    void test2() throws Exception {
      var raw = "{\"-ipglist\":\"hi\"}";
      var tgt = json.readValue(raw, configClass());

      assertEquals("hi", tgt.getIPGList());
    }
  }

  @Nested
  @DisplayName(Flag.NegativeIPGList)
  class NegativeIPGList1
  {
    @Test
    @DisplayName("serialization")
    void test1() throws Exception {
      var tgt = newConfig();
      tgt.setNegativeIPGList("hi");

      assertEquals("{\"-negative_ipglist\":\"hi\"}", json.writeValueAsString(tgt));
    }

    @Test
    @DisplayName("deserialization")
    void test2() throws Exception {
      var raw = "{\"-negative_ipglist\":\"hi\"}";
      var tgt = json.readValue(raw, configClass());

      assertEquals("hi", tgt.getNegativeIPGList());
    }
  }
}

File: src/test/java/org/veupathdb/lib/blast/BlastWithIPGListContractTest.kt
package org.veupathdb.lib.blast

import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.veupathdb.lib.blast.consts.Flag
import kotlin.reflect.KClass

@DisplayName("BlastWithIPGList")
internal abstract class BlastWithIPGListContractTest : BlastWithListsContractTest() {
  private val json = ObjectMapper()
  abstract override fun newConfig(): BlastWithIPGList
  abstract override fun configClass(): KClass<out BlastWithIPGList>

  @Nested
  @DisplayName(Flag.IPGList)
  internal inner class IPGList1 {
    @Test
    @DisplayName("serialization")
    @Throws(Exception::class)
    fun test1() {
      val tgt = newConfig()
      tgt.ipgList = "hi"
      Assertions.assertEquals("{\"-ipglist\":\"hi\"}", json.writeValueAsString(tgt))
    }

    @Test
    @DisplayName("deserialization")
    @Throws(Exception::class)
    fun test2() {
      val raw = "{\"-ipglist\":\"hi\"}"
      val tgt = json.readValue(raw, configClass().java)
      Assertions.assertEquals("hi", tgt.ipgList)
    }
  }

  @Nested
  @DisplayName(Flag.NegativeIPGList)
  internal inner class NegativeIPGList1 {
    @Test
    @DisplayName("serialization")
    @Throws(Exception::class)
    fun test1() {
      val tgt = newConfig()
      tgt.negativeIPGList = "hi"
      Assertions.assertEquals("{\"-negative_ipglist\":\"hi\"}", json.writeValueAsString(tgt))
    }

    @Test
    @DisplayName("deserialization")
    @Throws(Exception::class)
    fun test2() {
      val raw = "{\"-negative_ipglist\":\"hi\"}"
      val tgt = json.readValue(raw, configClass().java)
      Assertions.assertEquals("hi", tgt.negativeIPGList)
    }
  }
}